federation/scim/src/main/java/org/keycloak/federation/scim/core/service/UserScimService.java

```diff
@@ -0,0 +1,130 @@
+import org.keycloak.models.RoleModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.federation.scim.core.ScrimEndPointConfiguration;
+import org.keycloak.federation.scim.core.exceptions.InconsistentScimMappingException;
+import org.keycloak.federation.scim.core.exceptions.SkipOrStopStrategy;
+import org.keycloak.federation.scim.core.exceptions.UnexpectedScimDataException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class UserScimService extends AbstractScimService<UserModel, User> {
+    private static final Logger LOGGER = Logger.getLogger(UserScimService.class);
+
+    public UserScimService(KeycloakSession keycloakSession, ScrimEndPointConfiguration scimProviderConfiguration,
+            SkipOrStopStrategy skipOrStopStrategy) {
+        super(keycloakSession, scimProviderConfiguration, ScimResourceType.USER, skipOrStopStrategy);
+    }
+
+    @Override
+    protected Stream<UserModel> getResourceStream() {
+        return getKeycloakDao().getUsersStream();
+    }
+
+    @Override
+    protected boolean entityExists(KeycloakId keycloakId) {
+        return getKeycloakDao().userExists(keycloakId);
+    }
+
+    @Override
+    protected Optional<KeycloakId> matchKeycloakMappingByScimProperties(User resource) throws InconsistentScimMappingException {
+        Optional<KeycloakId> matchedByUsername = resource.getUserName().map(getKeycloakDao()::getUserByUsername)
+                .map(this::getId);
+        Optional<KeycloakId> matchedByEmail = resource.getEmails().stream().findFirst().flatMap(MultiComplexNode::getValue)
+                .map(getKeycloakDao()::getUserByEmail).map(this::getId);
+        if (matchedByUsername.isPresent() && matchedByEmail.isPresent() && !matchedByUsername.equals(matchedByEmail)) {
+            String inconstencyErrorMessage = "Found 2 possible users for remote user " + matchedByUsername.get() + " - "
+                    + matchedByEmail.get();
+            LOGGER.warn(inconstencyErrorMessage);
+            throw new InconsistentScimMappingException(inconstencyErrorMessage);
+        }
+        if (matchedByUsername.isPresent()) {
+            return matchedByUsername;
+        }
+        return matchedByEmail;
+    }
+
+    @Override
+    protected KeycloakId createEntity(User resource) throws UnexpectedScimDataException {
+        String username = resource.getUserName().filter(StringUtils::isNotBlank)
+                .orElseThrow(() -> new UnexpectedScimDataException(
+                        "Remote Scim user has empty username, can't create. Resource id = %s".formatted(resource.getId())));
+        UserModel user = getKeycloakDao().addUser(username);
+        resource.getEmails().stream().findFirst().flatMap(MultiComplexNode::getValue).ifPresent(user::setEmail);
+        boolean userEnabled = resource.isActive().orElse(false);
+        user.setEnabled(userEnabled);
```

Comment:
We should create users through the `UserProfileProvider` in order to enforce realm-level settings when managing users in a realm. For instance, the validations defined for user attributes in the `User Profile` tab in the `Realm Settings`.

Perhaps it also makes sense to allow pulling users without any validation so that they can update their profiles later. For that, we might want a switch.

For now, I don't see it as a blocker but looking at this code it seems we are only setting the email from the response from the SCIM server?